### DISCUSSION TITLE
Unable to invoke no-args constructor for interface com.slack.eithernet.ApiResult.

### DISCUSSION THREAD

**Discussion by mohitb117** [slackhq/EitherNet#38] [Q&A] [answered]
First of all, really really cool and clean application of sealed classes + retrofit responses 💯💯

While i was trying to integrate the library into into my test project,  I received an error : 
```
Error Received Unable to invoke no-args constructor for interface com.slack.eithernet.ApiResult. 
Registering an InstanceCreator with Gson for this type may fix this problem.
```

Did i miss a step in configuring my migration from retrofit Responses into ApiResult? 
Here is a [diff](https://github.com/mohitb117/OMDB_API/commit/f46da1f42ce6fe6a67ff72989743c779d0a57cbd) of the steps i took to migrate. 

Thanks again and any insight is greatly appreciated 🙏🏽

-----

Some more observations: 

Setup 1: No GsonConverterFactory as part of retrofit builder

```
val api = Retrofit.Builder()
  .addConverterFactory(ApiResultConverterFactory)
  .addCallAdapterFactory(ApiResultCallAdapterFactory)
  .build()
  .create<TestApi>()
```
Error: Error Received Unable to create converter for interface com.slack.eithernet.ApiResult
        for method TestApi.loadResults

Setup 2: With ApiResultConverterFactory after GsonConverterFactory
```
val api = Retrofit.Builder()
  .addConverterFactory(GsonConverterFactory.create())
  .addConverterFactory(ApiResultConverterFactory)
  .addCallAdapterFactory(ApiResultCallAdapterFactory)
  .build()
  .create<TestApi>()
```

Error: Error Received Unable to invoke no-args constructor for interface com.slack.eithernet.ApiResult. Registering an InstanceCreator with Gson for this type may fix this problem.

Setup 3: With GsonConverterFactory after  ApiResultConverterFactory
```
val api = Retrofit.Builder()
  .addConverterFactory(ApiResultConverterFactory)
  .addCallAdapterFactory(ApiResultCallAdapterFactory)
  .addConverterFactory(GsonConverterFactory.create())
  .build()
  .create<TestApi>()
```

Error: Error Received Unable to create converter for interface com.slack.eithernet.ApiResult
        for method TestApi.loadResults

**Comment by Shayne3000** [comment]
Hey man I came across something that might help you. One option you have is to hash it out and try and make it work with EitherNet. 

Another option is to build a custom solution yourself. Whilst it may seem daunting, if you read the article I linked below, it could help in demystifying the process. For your purposes, you might just need something that simple. 

Something I found that helped me truly understand the inner workings of Retrofit and its Network response handling behaviour was [this well-written article](https://proandroiddev.com/create-retrofit-calladapter-for-coroutines-to-handle-response-as-states-c102440de37a).

That alternative allows you to use Retrofit's Moshi/Gson converter and a custom CallAdapter + CallAdapterFactory. 

What I gleaned from the article were 4 steps to doing this.

```
Steps for handling Retrofit network responses using a Custom CallAdapter.
    1. Create a Sealed class that represents network call response states.
    2. Create a custom override of the Call interface to parameterize/convert network responses into
    the Sealed class.
    3. Create a custom CallAdapter class that takes the success response Type parameterized in a Call type
    and injects it into and delegates to the Custom Call interface we created a priori.
    4. Create a custom CallAdapter.Factory that initialises and returns the CallAdapter for network response handling
    if the return type of the api interface method is one that said CallAdapter can handle i.e. ApiResponse.
    If the return type of the api interface method is not one it can handle, null is returned as this isn't the right adapter and
    the factory delegates to an adapter (if present) that can handle that return type.
```

I hope this helps.

**Reply by mohitb117** [reply]
That is a good idea and i think i might end up exploring that on a separate thread but this library is a good vehicle to drive consistency and changes which is why i was hoping to make EitherNet work :)

**Comment by mohitb117** [comment, marked as answer]
Aah i figured out my problem: 
```
 ApiResult<SearchResultsBody, Nothing>
vs
ApiResult<SearchResultsBody, String>
```

and now it works 😅 , it seems i have to read a bit more on the Parsing spec 😄
